python/ray/experimental/serve/queues.py

```diff
@@ -220,3 +229,167 @@ def flush(self):
+                    backend_names, p=backend_weights).squeeze()
+
+                request = queue.popleft()
+                self.buffer_queues[chosen_backend].add(request)
+
+
+@ray.remote
+class RandomPolicyQueueActor(RandomPolicyQueue, CentralizedQueuesActor):
+    pass
+
+
+class RoundRobinPolicyQueue(CentralizedQueues):
+    """
+    A wrapper class for RoundRobin policy.
+    This backend selection policy is `Stateful` meaning
+    the current decisions of selecting backend are
+    dependent on previous decisions.
+    """
+
+    # Saves the information about last assigned
+    # backend for every service
+    round_robin_iter = {}
+
+    def set_traffic(self, service, traffic_dict):
+        logger.debug("Setting traffic for service %s to %s", service,
+                     traffic_dict)
+        self.traffic[service] = traffic_dict
+        backend_names = list(self.traffic[service].keys())
+        self.round_robin_iter[service] = itertools.cycle(backend_names)
+        self.flush()
+
+    def _flush_service_queue(self):
+        # perform traffic splitting for requests
+        for service, queue in self.queues.items():
+            # if there are incoming requests and there are backends
+            if len(queue) and len(self.traffic[service]):
+                if service not in self.round_robin_iter:
+                    backend_names = list(self.traffic[service].keys())
+                    self.round_robin_iter[service] = itertools.cycle(
+                        backend_names)
+
+                while len(queue):
+                    # choose the next backend available from persistent
+                    # information
+                    chosen_backend = next(self.round_robin_iter[service])
+                    request = queue.popleft()
+                    self.buffer_queues[chosen_backend].add(request)
+
+
+@ray.remote
+class RoundRobinPolicyQueueActor(RoundRobinPolicyQueue,
+                                 CentralizedQueuesActor):
+    pass
+
+
+class PowerOfTwoPolicyQueue(CentralizedQueues):
+    """
+    A wrapper class for powerOfTwo policy.
+    This backend selection policy is `Stateless` meaning
+    the current decisions of selecting backend are
+    dependent on previous decisions.
+    """
+
+    def _flush_service_queue(self):
+        # perform traffic splitting for requests
+        for service, queue in self.queues.items():
+            # while there are incoming requests and there are backends
+            while len(queue) and len(self.traffic[service]):
+                backend_names = list(self.traffic[service].keys())
+                backend_weights = list(self.traffic[service].values())
+                if len(self.traffic[service]) >= 2:
+                    # randomly pick 2 backends
+                    backend1, backend2 = np.random.choice(
+                        backend_names, 2, p=backend_weights)
+                    backend1, backend2 = backend1.squeeze(), backend2.squeeze()
```

Comment:
do we need to squeeze here? i though np.random.choice returns 1 dimensional vector?

Reply:
Yes this can be removed. I just emulated the previous code